### DISCUSSION TITLE
curl: (7) Failed to connect to more.musl.cc port 443: Connection timed out

### DISCUSSION THREAD

**Discussion by masnagam** [community/community#27906] [Actions] [answered]
Hi,

I'm facing an issue that `curl` fails:
https://github.com/mirakc/mirakc/runs/7612735719?check_suite_focus=true#step:8:2450

The execution of the workflow was scheduled by `on.schedule.cron`:
https://github.com/mirakc/mirakc/blob/main/.github/workflows/daily-main.yml#L7

and ran on `ubuntu-latest`.

A PAT was used in order to trigger other workflows:
https://github.com/mirakc/mirakc/blob/main/.github/workflows/daily-main.yml#L12

The `docker` job tried to build a Docker image by using `docker/build-push-action@v3`.
[buildenv.sh](https://github.com/mirakc/mirakc/blob/main/docker/build-scripts/buildenv.sh) was executed and tried to download a toolchain from `more.musl.cc` by using `curl`, but the connection timed out.

You can see verbose logs from `curl` in the following page:
https://github.com/mirakc/mirakc/runs/7598905955?check_suite_focus=true#step:8:8956

I confirmed that:

* [Another workflow](https://github.com/mirakc/mirakc/actions/runs/2762828922) built Docker images successfully, which was triggered by a push event
* Docker images can be built successfully on my local Linux PC by using a command described in [README.md](https://github.com/mirakc/mirakc/blob/main/docker/README.md#how-to-build-images-from-source)

Are there any mistakes in my scripts and/or the workflow definition?

Thank you for your help.

**Comment by gabek** [comment]
Same issue here.

```
      +crosscompiler | Connecting to musl.cc (musl.cc)|104.232.42.245|:443... failed: Operation timed out.
```

Some detail: It's not specifically curl, it's any connection to that host. For a few days it was failing intermittently but now it's consistently failing.

**Reply by zv-io** [reply]
For clarity, the reason it was failing "intermittently" is because I was shutting the web server down during huge traffic spikes (see https://github.com/orgs/community/discussions/27906#discussioncomment-3332440) just for a few minutes at a time to let the network recover.

Then, after this occurred several days in a row and again a week later, I blocked the Microsoft IP ranges.

This pattern should be rewritten: https://github.com/mirakc/mirakc/blob/b2d25a350da3877e511876c62f774f1b39a6e869/docker/build-scripts/buildenv.sh#L17-L20

**Reply by masnagam** [reply]
@zv-io
Thank you for your suggestion.  I'll improve the script.
Currently, I'm considering to use a pre-built Docker image as a build environment which contains toolchains needed for cross compilations.

**Reply by runlevel5** [reply]
@masnagam using pre-built Docker image like `taiki-e/install-action@cross` GitHub Action is a good option HOWEVER this solution does not cover all cross-compilation target such as `powerpc64le-unknown-linux-musl`

@zv-io just wondering if you could recommend any mirror that is not blocked by GitHub? Many thanks

**Reply by zv-io** [reply]
@masnagam I'd be more willing to set aside time to update the toolchains as well as set up mirrors if more than zero companies were willing to offer more than nothing in return for these. See my comment on the selected answer.

There is also a security risk: most people do not verify that the binaries they download match the expected checksums, so allowing third-party mirrors must be explicitly separate from the main site. I'd prefer to operate the infrastructure myself and I cannot in good faith keep track of or recommend any third-party mirrors.

**Reply by masnagam** [reply]
@zv-io 
i think you replied to @runlevel5 but our project no longer uses the toolchains downloaded from `musl.cc`.
thank you for your efforts so far.

@runlevel5 
thank you for sharing your knowledge.
our project actually uses `taiki-e/setup-cross-toolchain-action` for cross compilation tests.

**Comment by airtower-luna** [comment]
Timeout without any connection makes me think of some kind of firewall, that's usually the cause of something like this. Your workflow seems to be downloading other data just fine, so it's probably not the Actions infrastructure in general.

Theoretically it could be anywhere along the route (I've seen issues with firewalls eating "fragmentation needed" packets), but I'm wondering if maybe there's some kind of rate-limiting involved that's enforced by firewall. I wonder if you could contact the admins of `musl.cc` and ask if they have any such rules, and if yes if there's any overlap between them and the [IP ranges the GitHub API reports](https://docs.github.com/en/rest/meta) for Actions? :sweat_smile:

**Reply by gabek** [reply]
I went ahead and emailed musl.cc and I'll report if I hear back!

**Reply by masnagam** [reply]
@airtower-luna @gabek Thank you for your help :)

**Comment by jfullstackdev** [comment]
hey guys, I am using cURL directly on GitHub Codespaces without anything being installed ! ( You also have there by default Git, Bash, Java etc. ) That is awesome ! and unlike local installation like this that can interfere with the functionalities, you don't worry about that using GitHub Codespaces, 

anyways, it's a remote Linux machine but it's really that convenient

**Reply by gabek** [reply]
I'm not terribly knowledgeable about codespaces, but could you clarify how that is related to our GitHub actions issue? Thanks for chiming in!

**Reply by jfullstackdev** [reply]
I just saw above about the firewall and local setup that might interfere

that's why I mentioned about cURL and Codespaces

can you run that command separately ? manually doing in cURL ? just to be sure whehter it's working or not

**Comment by zv-io** [comment, marked as answer]
Hi, @gabek did email me. Thanks for reaching out and creating a discussion about this.

Currently, most Microsoft IP ranges are blocked from accessing the site. This is due to some careless, though probably unintentional, DDoS abuse. I'm happy to unblock these ranges but it's going to have to wait until September 01 for unrelated reasons.

At times, up to ~2k IP addresses had been hammering the site *at the same time* downloading the x86_64 and aarch64 toolchains *over and over*, all reporting some form of the `curl` user agent. I think so far approximately 30k unique Microsoft IP addresses had been used.

It wasn't clear to me initially whether this was a Microsoft-internal service deployment, some private Azure accounts, or GitHub Actions.

On my end, not just my site, but the upstream ISP, went offline or had service degradation that affected VoIP connections. This could potentially affect access to emergency services, so we've shaped our QoS policy to at least prevent that.

So I wholesale blocked most of the Microsoft IP ranges I could find. Sorry.

My position is this: if these toolchains (and maintenance, including new releases, more testing, etc.) are important to the GitHub community, I'd be happy to work out some sort of sponsored arrangement to prioritize this work, have more robust hosting (CDN or otherwise), and allow everyone to access these toolchains at full speed (10Gbps instead of 100Mbps).

There is absolutely no reason that these toolchains need to be downloaded from the upstream server every single time CI is run. Let's not forget that bandwidth isn't unlimited, free, etc.

**Reply by gabek** [reply]
This makes a lot of sense, thanks for clarifying. Personally I was downloading once a day during a nightly build not knowing I was adding to a growing problem for you. I'll certainly make changes to not have to rely on your server anymore.

**Reply by joeyparrish** [reply]
May I suggest mirroring those toolchains on GitHub itself?  Then you can require that GitHub Actions users download toolchains from inside GitHub.  You could do this using GitHub releases or as files in a repo or using GitHub Pages.  Make GitHub be your CDN!  I would be happy to help build automation for this if it's useful.

**Reply by joeyparrish** [reply]
Actually, GitHub Pages has a 1GB limit per site, and you appear to have about 9GB of toolchains hosted.

I managed to offload the two toolchains I need for my project into a personal GitHub Pages site, but it doesn't appear to be a good solution for everyone and every toolchain.

@zv-io, I'm motivated to find a solution for this.  Please reach out to me at the email address on my GitHub profile if you want to work together.  If I know how much traffic we could expect, I could go pitch my company on hosting or gifting cloud storage credits.

**Reply by zv-io** [reply]
> @zv-io, I'm motivated to find a solution for this. Please reach out to me at the email address on my GitHub profile if you want to work together. If I know how much traffic we could expect, I could go pitch my company on hosting or gifting cloud storage credits.

Thank you; I appreciate the offer. I will reach out from the same email in my profile (so no one poses as being me).

I had been previously talking to a popular CDN-type company but they didn't seem to be interested in helping unless it was paired with a branding opportunity. Obviously a non-starter.

Regarding storage requirements, I am quoting from an email I sent to them the other month:

```
As above, just note that storage requirements are probably the
biggest concern. It's about ~100MB per toolchain. There ideally
will be a 60x60 matrix of "from" and "to" targets, with "native"
being the diagonal of the matrix, where from=to. For Linux.

I just need time to work on that; have been meaning to finish.

~350GB per GCC version, maybe 10 versions (5-15 so far), so
that's 3.5TB for Linux toolchains. Divide by 2, approximately,
because GCC doesn't build for all those targets on all versions,
so conservatively we're at 2TB for Linux toolchains.

For Windows, it is a bit simpler, just from 32- and 64- bit, to
60 targets. That's 6GB per version, or 60GB for all versions.

For macOS, it's about the same, but PowerPC + Intel + AppleSi,
for each of 60 targets. That's 18GB per version, or 180GB for
all versions.

For Solaris, it's like Windows, so 60GB.

Add all those up, now we're about 2.5TB conservatively.

That leaves some space for experiments/testing/prototypes/etc.

Bandwidth has been about 1TB/day to GitHub, which isn't that
much in terms of tech, but it strains a network that shares
bandwidth between a bunch of other FOSS projects.

So that's approximately what can be expected of a mirror.
```

**Reply by ZeroAurora** [reply]
> Actually, GitHub Pages has a 1GB limit per site, and you appear to have about 9GB of toolchains hosted.

GitHub Releases don't have such limitations
